include/beman/execution26/detail/as_awaitable.hpp

```diff
@@ -14,13 +22,22 @@ namespace beman::execution26 {
  * \headerfile beman/execution26/execution.hpp <beman/execution26/execution.hpp>
  */
 struct as_awaitable_t {
-    template <typename E, typename P>
-    auto operator()(E&& e, P& p) const {
-        if constexpr (requires { ::std::forward<E>(e).as_awaitable(p); }) {
-            //-dk:TODO mandates is-awaitable<E, Promise>
-            return ::std::forward<E>(e).as_awaitable(p);
+    template <typename Expr, typename Promise>
+    auto operator()(Expr&& expr, Promise& promise) const {
+        if constexpr (requires { ::std::forward<Expr>(expr).as_awaitable(promise); }) {
+            static_assert(
+                ::beman::execution26::detail::is_awaitable<decltype(::std::forward<Expr>(expr).as_awaitable(promise)),
+                                                           Promise>,
+                "as_awaitable must return an awaitable");
+            return ::std::forward<Expr>(expr).as_awaitable(promise);
+        } else if constexpr (::beman::execution26::detail::
+                                 is_awaitable<Expr, ::beman::execution26::detail::unspecified_promise>) {
+            return (void(promise), ::std::forward<Expr>(expr));
+        } else if constexpr (::beman::execution26::detail::awaitable_sender<Expr, Promise>) {
+            return ::beman::execution26::detail::sender_awaitable<Expr, Promise>{::std::forward<Expr>(expr), promise};
+        } else {
+            return (void(promise), ::std::forward<Expr>(expr));
```

Comment:
`void(promise)` doesn't do anything. It is in the specification because `as_awaitable(e, p)` is described as _expression equivalent to_ and without `void(p)` the expression `p` isn't evaluated. However, when calling `as_awaitable_t::operator()` the expression is clearly evaluated and naming it doesn't really do anything.

Reply:
changed